otter/convergence/gathering.py

```diff
@@ -112,8 +116,71 @@ def group_id(s):
                             filter(server_predicate),
                             filter(has_group_id))
 
-    eff = get_all_server_details(changes_since)
-    return eff.on(servers_apply)
+    return get_all_server_details(changes_since).on(servers_apply)
+
+
+def merge_servers(first, second):
+    """
+    Return servers got by merging the two sets of servers where second
+    takes precedence over first
+
+    :param list first: List of server dicts
+    :param list second: List of server dicts
+    :return: List of merged server dicts
+    """
+    def sdict(servers):
+        return {s['id']: s for s in servers}
+
+    return merge(sdict(first), sdict(second)).values()
+
+
+def server_of_group(group_id, server):
+    """
+    Return True if server belongs to group_id. False otherwise
+    """
+    return group_id_from_metadata(server.get('metadata', {})) == group_id
+
+
+def group_servers(group_id, grouped_servers):
+    """
+    Return servers of a group defaulting to empty if not found
+    """
+    return grouped_servers.get(group_id, [])
+
+
+@do
+def get_scaling_group_servers(tenant_id, group_id, now,
+                              cache_class=CassScalingGroupServersCache,
+                              all_as_servers=get_all_scaling_group_servers,
+                              all_servers=get_all_server_details):
+    """
+    Get a group's servers taken from cache if it exists. Updates cache
+    if it is empty from newly fetched servers
+    # NOTE: This function takes tenant_id even though the whole effect is
+    # scoped on the tenant because cache calls require tenant_id. Should
+    # they also not take tenant_id and work on the scope?
```

Comment:
We could perhaps treat TenantScope more like a reader monad, and create a "GetTenantID" effect.

Which then really begs for a general DynamicScope effect with a GetVariable effect. Oh man, I feel dirty. :) And it means that Effect could implement the vision of twisted.python.context in a coherent way! I would feel a lot less dirty about this idea if we had a type system that could ensure all dynamic accesses were valid. 

We could still do this without completely generalizing to dynamic scoping by making TenantScope provide a performer for an `UpdateServersCache` intent, just  like it does for `ServiceRequest`.

Anyway, this is just some bikeshedding. I don't think you need to change this code.